module/zfs/zil.c

```diff
@@ -3140,18 +3105,14 @@ zil_commit_waiter_timeout(zilog_t *zilog, zil_commit_waiter_t *zcw)
 		 *   because it's blocked trying to acquire the waiter's
 		 *   lock, which occurs prior to calling dmu_tx_commit()
 		 */
-		mutex_exit(&zcw->zcw_lock);
 		zil_lwb_write_issue(zilog, lwb);
 		lwb = NULL;
 		zil_commit_writer_stall(zilog);
-		mutex_enter(&zcw->zcw_lock);
 	}
-
-out:
 	mutex_exit(&zilog->zl_issuer_lock);
 	if (lwb)
 		zil_lwb_write_issue(zilog, lwb);
-	ASSERT(MUTEX_HELD(&zcw->zcw_lock));
+	mutex_enter(&zcw->zcw_lock);
```

Comment:
Consider restructuring this code to avoid using`lwb == NULL` as a "flag":
```
    zil_lwb_write_issue(zilog, lwb);
    zil_commit_writer_stall(zilog);
    mutex_exit(&zilog->zl_issuer_lock);
} else {
    mutex_exit(&zilog->zl_issuer_lock);
    zil_lwb_write_issue(zilog, lwb);
}
mutex_enter(&zcw->zcw_lock);
```

Reply:
Done.